### ISSUE TITLE
In my model, GN+WS is worse

### ISSUE THREAD

**Issue by DrewdropLife** [joe-siyuan-qiao/WeightStandardization#6]
Hello, I am very interested in your method. After I added the WS method to my network (I replaced all the convolution layers followed by BN,  including GlobalConvBlock and ResidualBlock, and replaced all BN with GN), I found that I had to choose a smaller learning rate (reduced by 50 times, now is 2e-5) , because the original learning rate would cause huge losses and make training impossible. After a certain number of iterations, the results are not impressive, IOU has decreased, and the loss function curve is more turbulent than before. What did I do wrong? 

The framework I use is PyTorch, but I'm not familiar with it.Thank you for your advice!

The following code is the modified part of the GlobalConvBlock:

> class GlobalConvBlock(nn.Module):

>     def __init__(self, in_dim, out_dim, kernel_size):
>         super(GlobalConvBlock, self).__init__()
>         pad0 = (kernel_size[0] - 1) / 2
>         pad1 = (kernel_size[1] - 1) / 2
>         self.conv_l1 = nn.Conv2d(in_dim, out_dim, kernel_size=(kernel_size[0], 1),padding=(pad0, 0))
>         self.conv_l2 = Conv2d(out_dim, out_dim, kernel_size=(1, kernel_size[1]),padding=(0, pad1))#replace
>         self.conv_r1 = nn.Conv2d(in_dim, out_dim, kernel_size=(1, kernel_size[1]),padding=(0, pad1))
>         self.conv_r2 = Conv2d(out_dim, out_dim, kernel_size=(kernel_size[0], 1),padding=(pad0, 0))#replace
>         for m in self.modules():
>             if isinstance(m, nn.Conv2d):
>                 n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
>                 m.weight.data.normal_(0, sqrt(2. / n))
>                 if m.bias is not None:
>                     m.bias.data.zero_()
>             elif isinstance(m, Conv2d):             #I also added this initialization, right?
>                 n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
>                 m.weight.data.normal_(0, sqrt(2. / n))
>                 if m.bias is not None:
>                     m.bias.data.zero_()
>             elif isinstance(m, nn.GroupNorm):
>                 m.weight.data.normal_(1.0, 0.02)
>                 m.bias.data.fill_(0)
>     def forward(self, x):
>         x_l = self.conv_l1(x)
>         x_l = self.conv_l2(x_l)
>         x_r = self.conv_r1(x)
>         x_r = self.conv_r2(x_r)
>         #combine two paths
>         x = x_l + x_r
>         return x

This is the parameter set in my code layer:

>         self.conv1 = Conv2d(indim, indim*2, kernel_size=1, bias=False)
>         self.norm1 = nn.GroupNorm(num_groups = 32,num_channels = indim*2)
>         self.relu1 = nn.LeakyReLU(0.2, inplace=True)
>         # ……
>         self.deconvblock4 = nn.Sequential(
>             GlobalConvBlock(ndf*8*2, ndf*8, (7, 7)),
>             nn.GroupNorm(num_groups = 32,num_channels = ndf * 8),
>             nn.ReLU(True),
>         )

Thank you!

**Comment by joe-siyuan-qiao** [comment]
Thanks for your interest in WS.

It's quite difficult to diagnose the problems given the limited knowledge of your system so I will try to make some comments and hope they can be helpful.

1. It seems that conv_l1 and conv_r1 should also use WS and maybe turn their (including l2 and r2) bias term off.
2. If your system involves pretraining on large datasets, make sure you also pretrain a WS enabled backbone.
3. The learning rates for BN also work well for GN+WS on ImageNet, so 50x learning rate deduction for your problem is a little bit too much. There are some things you can check:
    a. Check if BN+WS works. If it does, then consider changing the hyper-parameters for GN.
    b. Check if learning rate warmup works. The performance drop might be due to small learning rates.
    c. Check if your final output uses WS (which it shouldn't). It can cause learning problem.

You can also share more information so we can provide more specific comments.

Thanks.

**Comment by DrewdropLife** [comment]
Thank you very much for your comments! I will check and revise it in turn according to your advice.

In addition, my whole model is trained from random initialization, without using pre-training model or dataset. I will check the impact of learning rate on different combinations of BN, GN and WS, and try to turn off bias items of all convolution layers.

Thank you again for taking the valuable time to give me advice!

**Comment by DrewdropLife** [comment]
Thank you again! I've found the problem. 
Convolution, normalization and activation are defined in residual blocks, but normalization is not used when forward. So I should use all nn.Conv2d, but I use WS convolution. This is my negligence, Now it has been corrected and good results have been achieved. You have helped me a lot! Thank you very much!

**Comment by joe-siyuan-qiao** [comment]
Great! Glad to hear that.
